src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Receivers.kt

```diff
@@ -140,17 +137,96 @@ open class TaskReceiver(tasksDirectory: String,
 class MarkdownTaskReceiver(
         private val texCompiler: CompilerApi,
         tasksDirectory: String,
-        resultPublisher: PublisherApi,
-        private val config: Config = DEFAULT_CONFIG) : TaskReceiver(tasksDirectory, resultPublisher) {
+        private val resultPublisher: PublisherApi,
+        private val executor: ExecutorService,
+        private val config: Config = DEFAULT_CONFIG) : CompilerFleetMessageReceiver() {
 
+    private val tasksDir: Path = getTaskDir(tasksDirectory)
     private val arguments = listOf("projectRootAbsPath", "workingDirRelPath",
             "inputFileName", "outputFileName", "mainFont")
     private val COMPILE_COMMAND_KEY = "pandoc.compile.command"
+    private val currentTasks = ConcurrentHashMap<String, Future<Boolean>>()
 
     override fun processMessage(message: PubsubMessage): Boolean {
-        val request = CompileRequest.parseFrom(message.data)
+        val request = Request.parseFrom(message.data)
+
+        when (request.typeCase) {
+            Request.TypeCase.COMPILE -> {
+                processCompile(request.compile)
+                return true
+            }
+
+            Request.TypeCase.CANCEL  -> {
+                return processCancel(request.cancel)
+            }
+
+            else -> {
+                LOGGER.error("Request type = {} is not set properly", request.typeCase)
+            }
+        }
+
+        return true
+    }
+
+    private fun processCancel(request: CancelRequestProto): Boolean {
+        LOGGER.debug("Canceling the task with id = {}", request.taskId)
+        val status = cancelTask(request)
+        val data = CompilerFleet.Cancel.newBuilder()
+                .setTaskId(request.taskId)
+                .setStatus(status)
+                .setCpuTime(0) // TODO: put value when cancel task is fixed
+                .build()
+                .toByteString()
+        val onPublishFailureCallback = {
+            LOGGER.info("Publish cancel task ${request.taskId} response failed with code ${StatusCode.FAILURE}")
+        }
+
+        resultPublisher.publish(data, onPublishFailureCallback)
```

Comment:
So this will publish serialized "Cancel", not serialized "CompilerFleetResult" with "Cancel" inside?